### DISCUSSION TITLE
Provide "if x := .." syntax also in for-loops

### DISCUSSION THREAD

**Discussion by jdonnerstag** [vlang/v#12673] [Ideas]
V supports `if` optionals unwrapping as described [here](https://github.com/vlang/v/blob/master/doc/docs.md#handling-optionals). I propse to provide something similar in `for` loops as well.

Reasoning: Consider how many of the `io.read...()` methods are implemented and how they shall be used:
```
for {
    line := br.read_line() or { break }
    // ...
}
```

With the proposed change, the following would be possible. IMHO more readable and also more obvious.
```
for line := br.read_line() { ... }
```

As long as read_line() is successful, the next line will be returned. Upon an error, the for-loop will break. It would probably be useful to distinguish between true errors and "reached the end". True errors may return from the function with that error ?!?

I envision that this change greatly simplifies creating iterators, and that it is a true improvement to V-lang. Currently a `next()?` function is needed. Which btw also returns an optional.

**Comment by dumblob** [comment]
Generally I find the idea OK.

>With the proposed change, the following would be possible. IMHO more readable and also more obvious.
>
>`for line := br.read_line() { ... }`

But this is IMHO totally wrong. It **hides** the fact that `read_line()` returns an optional (or maybe later a [`result` **and** `optional`](https://github.com/vlang/rfcs/pull/7 )) which is absolutely not acceptable. Feel free to amend the example to adhere to the transparency *V* ecosystem pushes for :wink:.

**Comment by jdonnerstag** [comment]
"It hides the fact that read_line() returns an optional (or maybe later a result and optional)" => True, it does. Very much like in `if x := my_fn_returning_optional() { .. }` does and also like `for x in my_data { ... }`  hides that a next() function is called, which returns an optional that determines whether the for loop breaks or not.  With your line of argument, both would need to be removed from V's syntax.

**Comment by dumblob** [comment]
>Very much like in `if x := my_fn_returning_optional() { .. }` does and also like `for x in my_data { ... }` hides that a next() function is called, which returns an optional that determines whether the for loop breaks or not. With your line of argument, both would need to be removed from V's syntax.

It's two distinct (unrelated) things. One is about error handling (or generally "sumtype handling") and the other is about a pure function (or generic).

So, `if x := my_fn_returning_optional() { .. }` won't compile. And if it will (I didn't test it), then it's a serious bug in *V* compiler - please report it then.

And `for x in my_data { ... }` "hiding" that a `next()` function is called is fine as the compiler checks for types etc. It's the same level of abstraction as wrapping a function call in another function. There is really no *hidden* behavior - it does exactly what the line says **without side effects** (unlike the proposal `for line := br.read_line() { ... }`).

**Comment by CoderCharmander** [comment]
> So, `if x := my_fn_returning_optional() { .. }` won't compile. And if it will (I didn't test it), then it's a serious bug in _V_ compiler - please report it then.

This syntax is a feature: the code inside the `if` runs *only* if the optional is not an error (or `none`), with the return value assigned to `x`.

**Comment by dumblob** [comment]
> > So, `if x := my_fn_returning_optional() { .. }` won't compile. And if it will (I didn't test it), then it's a serious bug in _V_ compiler - please report it then.
> 
> This syntax is a feature: the code inside the `if` runs _only_ if the optional is not an error (or `none`), with the return value assigned to `x`.

No way :cold_sweat:. I've tried it now and you're right :open_mouth:. This totally jeopardizes the whole [optionals & results endeavor](https://github.com/vlang/rfcs/pull/7 ).

@medvednikov could you chime in and explain why this hidden, unsafe, unintuitive, surprising, and **unnecessary** (what's wrong with `if x := my_fn_returning_optional()? { .. }`? - notice the `?` at the end of the call; for anything else there is `match` - remember "there is only one way of doing things") behavior and how should this be compatible (and consistent!) with the RFC https://github.com/vlang/rfcs/pull/7 ?
